tq/adapterbase.go

```diff
@@ -82,19 +90,58 @@ func (a *adapterBase) Begin(maxConcurrency int, cb TransferProgressCallback, com
 	return nil
 }
 
-func (a *adapterBase) Add(t *Transfer) {
-	tracerx.Printf("xfer: adapter %q Add() for %q", a.Name(), t.Object.Oid)
-	a.jobChan <- t
+type job struct {
+	T *Transfer
+
+	listeners []chan<- TransferResult
+	wg        *sync.WaitGroup
+}
+
+func (j *job) Done(err error) {
+	for _, l := range j.listeners {
+		l <- TransferResult{j.T, err}
+	}
+
+	j.wg.Done()
+}
+
+func (a *adapterBase) Add(transfers ...*Transfer) <-chan TransferResult {
+	results := make(chan TransferResult, len(transfers))
+
+	listeners := []chan<- TransferResult{results}
+	if a.outChan != nil {
+		listeners = append(listeners, a.outChan)
+	}
+
+	a.jobWait.Add(len(transfers))
+
+	go func() {
+		defer close(results)
+
+		for _, t := range transfers {
+			// BUG(taylor): End() is race-y here, and can close
+			// jobChan before we want it to
+			a.jobChan <- &job{t, listeners, a.jobWait}
```

Comment:
This is correct, if the calling code occurs in a single goroutine. Consider this:

```go
func main() {
        var a *adapterBase

        go a.Add(myBatch)

        a.Wait()
}
```

If we're say, 50 items into `myBatch` before the execution scheduler takes us to the `a.Wait()` call, it'll close `a.jobChan`, which is what we expect. Once the scheduler takes us back to the goroutine, it'll try and process the 51st item and write into a closed channel, which is a panic. Even if it weren't a panic, we don't want to lose those items.